service/matching/task_queue_manager.go

```diff
@@ -783,40 +813,42 @@ func (c *taskQueueManagerImpl) userDataFetchSource() (string, error) {
 func (c *taskQueueManagerImpl) fetchUserData(ctx context.Context) error {
 	ctx = c.callerInfoContext(ctx)
 
-	if !c.config.LoadUserData() {
-		// if disabled, mark ready now
-		c.SetUserDataInitialFetch(nil)
-		return nil
-	}
 	if c.managesSpecificVersionSet() {
 		// tqm for specific version set doesn't have its own user data
-		c.SetUserDataInitialFetch(nil)
+		c.SetUserDataStatus(userDataSpecificVersion, nil)
 		return nil
 	}
-	if c.db.DbStoresUserData() {
-		// root workflow partition "owns" user data, read it from db
-		err := c.db.loadUserData(ctx)
-		c.SetUserDataInitialFetch(err)
-		return err
-	}
 
 	// otherwise fetch from parent partition
 
 	fetchSource, err := c.userDataFetchSource()
 	if err != nil {
 		if err == errMissingNormalQueueName { // nolint:goerr113
-			// pretend we have no user data
-			c.SetUserDataInitialFetch(nil)
+			// pretend we have no user data. this is a sticky queue so the only effect is that we can't
+			// kick off versioned pollers.
+			c.SetUserDataStatus(userDataEnabled, nil)
 		}
 		return err
 	}
 
-	firstCall := true
+	// hasFetchedUserData is true if we have gotten a successful reply to GetTaskQueueUserData.
+	// It's used to control whether we do a long poll or a simple get.
+	hasFetchedUserData := false
 
 	op := func(ctx context.Context) error {
+		if !c.config.LoadUserData() {
+			// if disabled, mark disabled and ready, but allow retries so that we notice if
+			// it's re-enabled
+			c.SetUserDataStatus(userDataDisabled, nil)
+			return errUserDataDisabled
+		}
+
 		knownUserData, _, err := c.GetUserData(ctx)
 		if err != nil {
-			return err
+			// Start with a non-long poll after re-enabling after disable, so that we don't have to wait the
```

Comment:
The phrase, "non-long poll after re-enabling after disable," is pretty dense. Also, I'm not sure it's relevant. I think all we need to express here is that, if we have user data, we want to wait for changes to it; if we don't have any user data, we want to fetch the latest.

Reply:
sure, that's one way to look at it. I think the only subtle part is that we may have user data, then get disabled, then get re-enabled. in that case `hasFetchedUserData` would be true here, so we have to explicitly set it to false. I'm explaining why we explicitly set it to false